File: pidex/sources/network.py
import logging
import threading
import time
from datetime import datetime

from pidex.core.bus import EventBus
from pidex.core.constants import EVENT_INTERFACE_DOWN, EVENT_INTERFACE_UP, SEVERITY_WARN, SOURCE_NETWORK
from pidex.core.event import Event
from pidex.sources.base import BaseSource

logger = logging.getLogger(__name__)


class NetworkSource(BaseSource):
    def run(self, stop_event: threading.Event) -> None:
        try:
            from pyroute2 import IPRoute

            ipr = IPRoute()
        except Exception:
            logger.error("pyroute2 not available — skipping network source")
            return

        try:
            ipr.bind()
            while not stop_event.is_set():
                for msg in ipr.get():
                    event = _parse_netlink(msg)
                    if event is not None:
                        self._bus.publish(event)
        except Exception:
            logger.exception("Netlink error")
        finally:
            ipr.close()


def _parse_netlink(msg: dict) -> Event | None:
    event = msg.get("event")
    if event == "RTM_DELLINK":
        ifindex = msg.get("index")
        attrs = {k: v for k, v in msg.get("attrs", [])}
        name = attrs.get("IFLA_IFNAME", f"if{ifindex}")
        ts = _extract_ts(msg)
        return Event(
            source=SOURCE_NETWORK,
            event_type=EVENT_INTERFACE_DOWN,
            severity=SEVERITY_WARN,
            title="Interface Removed",
            message=f"{name} removed",
            timestamp=ts,
        )

    if event != "RTM_NEWLINK":
        return None

    ifindex = msg.get("index")
    attrs = {k: v for k, v in msg.get("attrs", [])}
    name = attrs.get("IFLA_IFNAME", f"if{ifindex}")
    operstate = attrs.get("IFLA_OPERSTATE", "unknown")
    ts = _extract_ts(msg)

    if operstate == "UP":
        return Event(
            source=SOURCE_NETWORK,
            event_type=EVENT_INTERFACE_UP,
            severity=SEVERITY_WARN,
            title="Interface Up",
            message=f"{name} is UP",
            timestamp=ts,
        )
    elif operstate == "DOWN":
        return Event(
            source=SOURCE_NETWORK,
            event_type=EVENT_INTERFACE_DOWN,
            severity=SEVERITY_WARN,
            title="Interface Down",
            message=f"{name} is DOWN",
            timestamp=ts,
        )

    return None


def _extract_ts(msg: dict) -> datetime:
    ts = msg.get("timestamp")
    if ts is None:
        return datetime.now()
    return datetime.fromtimestamp(ts)

File: internal/source/network.go
package source

import (
	"context"
	"encoding/binary"
	"fmt"
	"log"
	"syscall"
	"time"
	"unsafe"

	"github.com/Shivamingale3/pi_dex/internal/core"
)

type NetworkSource struct {
	bus *core.EventBus
}

func NewNetworkSource(bus *core.EventBus) *NetworkSource {
	return &NetworkSource{bus: bus}
}

func (s *NetworkSource) Run(ctx context.Context) error {
	fd, err := syscall.Socket(syscall.AF_NETLINK, syscall.SOCK_RAW, syscall.NETLINK_ROUTE)
	if err != nil {
		log.Printf("network: socket: %v", err)
		return err
	}
	defer syscall.Close(fd)

	addr := &syscall.SockaddrNetlink{
		Family: syscall.AF_NETLINK,
		Groups: 1, // RTMGRP_LINK
	}
	if err := syscall.Bind(fd, addr); err != nil {
		log.Printf("network: bind: %v", err)
		return err
	}

	buf := make([]byte, 4096)
	for {
		select {
		case <-ctx.Done():
			return nil
		default:
		}

		n, _, err := syscall.Recvfrom(fd, buf, 0)
		if err != nil {
			log.Printf("network: recv: %v", err)
			time.Sleep(time.Second)
			continue
		}

		msgs, err := syscall.ParseNetlinkMessage(buf[:n])
		if err != nil {
			continue
		}

		for _, msg := range msgs {
			event := s.parseLink(msg)
			if event != nil {
				s.bus.Publish(*event)
			}
		}
	}
}

type ifinfomsg struct {
	Family uint8
	_      uint8
	Type   uint16
	Index  int32
	Flags  uint32
	Change uint32
}

func (s *NetworkSource) parseLink(msg syscall.NetlinkMessage) *core.Event {
	if msg.Header.Type != syscall.RTM_NEWLINK && msg.Header.Type != syscall.RTM_DELLINK {
		return nil
	}

	if len(msg.Data) < int(unsafe.Sizeof(ifinfomsg{})) {
		return nil
	}

	ifim := (*ifinfomsg)(unsafe.Pointer(&msg.Data[0]))
	ifindex := ifim.Index
	attrs := parseRtAttrs(msg.Data[unsafe.Sizeof(ifinfomsg{}):])
	name, _ := attrs["IFLA_IFNAME"].(string)
	if name == "" {
		name = fmt.Sprintf("if%d", ifindex)
	}

	ts := time.Now()

	if msg.Header.Type == syscall.RTM_DELLINK {
		return &core.Event{
			Source:    core.SourceNetwork,
			EventType: core.EventInterfaceDown,
			Severity:  core.SeverityWarn,
			Title:     "Interface Removed",
			Message:   name + " removed",
			Timestamp: ts,
		}
	}

	operstate, _ := attrs["IFLA_OPERSTATE"].(string)
	switch operstate {
	case "UP":
		return &core.Event{
			Source:    core.SourceNetwork,
			EventType: core.EventInterfaceUp,
			Severity:  core.SeverityWarn,
			Title:     "Interface Up",
			Message:   name + " is UP",
			Timestamp: ts,
		}
	case "DOWN":
		return &core.Event{
			Source:    core.SourceNetwork,
			EventType: core.EventInterfaceDown,
			Severity:  core.SeverityWarn,
			Title:     "Interface Down",
			Message:   name + " is DOWN",
			Timestamp: ts,
		}
	}

	return nil
}

func parseRtAttrs(data []byte) map[string]any {
	attrs := make(map[string]any)
	for i := 0; i+4 <= len(data); {
		length := int(binary.LittleEndian.Uint16(data[i : i+2]))
		rtype := int(binary.LittleEndian.Uint16(data[i+2 : i+4]))
		if length < 4 {
			break
		}
		if i+length > len(data) {
			break
		}
		payload := data[i+4 : i+length]
		// Pad to 4 bytes
		_ = payload
		switch rtype {
		case 3: // IFLA_IFNAME
			attrs["IFLA_IFNAME"] = cString(payload)
		case 16: // IFLA_OPERSTATE
			if len(payload) > 0 {
				attrs["IFLA_OPERSTATE"] = cString(payload[:1])
			}
		}
		i += ralign(length)
	}
	return attrs
}

func ralign(n int) int {
	return (n + 3) & ^3
}

func cString(b []byte) string {
	for i, c := range b {
		if c == 0 {
			return string(b[:i])
		}
	}
	return string(b)
}
